File: src/loader.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <arpa/inet.h>
#include <bpf/libbpf.h>
#include <linux/if_link.h>
#include <linux/bpf.h>
#include <sys/resource.h>
#include <net/if.h>
#include "../include/omega.h"


#define INTERFACE "enp4s0"


// Improvement: Checking for signal and when program exists properly cleanup and detach from network interface


// Print MAC address
void print_mac(__u8 *mac) {
    printf("%02x:%02x:%02x:%02x:%02x:%02x",
           mac[0], mac[1], mac[2], mac[3], mac[4], mac[5]);
}

// Callback for received events
static int handle_evt(void *ctx, void *data, size_t sz) {
    const struct packet_evt *evt = data;

    fprintf(stdout, "Received Event: SRC IP: %u DST IP: %u SRC PORT: %u DST PORT: %u\n",
        evt->src_ip, evt->dst_ip, evt->src_port, evt->dst_port);
    fflush(stdout);

    return 0;
}

int main() {
    struct bpf_object *obj;
    struct ring_buffer *rb = NULL;
    int map_fd, prog_fd, ifindex;

    ifindex = if_nametoindex(INTERFACE);
    if (!ifindex) {
        perror("Failed to get interface index");
        return 1;
    }
	printf("IF INDEX:%d\n", ifindex);

    obj = bpf_object__open_file("omega.bpf.o", NULL);
    if (!obj) {
        perror("Failed to open eBPF object file");
        return 1;
    }

// Load the BPF program into the kernel
if (bpf_object__load(obj)) {
    perror("Failed to load eBPF object");
    return 1;
}

// Debug: List all programs
struct bpf_program *prog;
bpf_object__for_each_program(prog, obj) {
    printf("Found BPF program: %s\n", bpf_program__name(prog));
}


    prog_fd = bpf_program__fd(bpf_object__find_program_by_name(obj, "receive_egress"));
    if (prog_fd < 0) {
        perror("Failed to find BPF program");
        return 1;
    }

    if (bpf_xdp_attach(ifindex, prog_fd, XDP_FLAGS_UPDATE_IF_NOEXIST, NULL) < 0) {
        perror("Failed to attach XDP program");
        return 1;
    }

    map_fd = bpf_map__fd(bpf_object__find_map_by_name(obj, "rb"));
    if (map_fd < 0) {
        perror("Failed to find BPF ring buffer map");
        return 1;
    }

    rb = ring_buffer__new(map_fd, handle_evt, NULL, NULL);
    if (!rb) {
        perror("Failed to create ring buffer");
        return 1;
    }

    printf("Listening for packets on %s...\n", INTERFACE);

    while (1) {
        ring_buffer__poll(rb, 100);
    }

    return 0;
}

File: src/loader.go
package main

import (
	"fmt"
	"net"
	"log"
	"time"
    "os"
    "os/signal"
    "syscall"
	"bytes"
	"encoding/binary"

	"github.com/cilium/ebpf"
	"github.com/cilium/ebpf/link"
	"github.com/cilium/ebpf/ringbuf"
	"github.com/cilium/ebpf/rlimit"

)


const (
	interfaceName = "enp4s0"
	bpfObjectName = "omega.bpf.o"
	mapName = "rb"
	functionName = "receive_egress"
)


type PacketEvt struct {
	SrcIP   uint32
	DestIP   uint32
	SrcPort uint16
	DestPort uint16
}

type Event struct {
	Source string
	Destination string
	Timestamp string
}


func formatIpPort(ip uint32, port uint16) string {
	return fmt.Sprintf("%d.%d.%d.%d:%d", byte(ip), byte(ip>>8), byte(ip>>16), byte(ip>>24), port)
}


func main() {
	stopper := make(chan os.Signal, 1)
	signal.Notify(stopper, os.Interrupt, syscall.SIGTERM)

	if err := rlimit.RemoveMemlock(); err != nil {
		panic(err)
	}

	obj, err := ebpf.LoadCollection(bpfObjectName)
	if err != nil {
		panic(fmt.Sprintf("Failed to load eBPF object: %v", err))
	}
	defer obj.Close()

	prog, found := obj.Programs[functionName]
	if !found {
		panic("Failed to find eBPF program")
	}

    
    iface_idx, err := net.InterfaceByName(interfaceName)
    if err != nil {
		panic(fmt.Sprintf("Failed to get interface %s: %v\n", interfaceName, err));
    }

    opts := link.XDPOptions{
		Program:   prog,
        Interface: iface_idx.Index,
    }

    lnk, err := link.AttachXDP(opts)
    if err != nil {
		panic(err)
    }
    defer lnk.Close()

    fmt.Println("Successfully loaded and attached BPF program.");

	ringBuf, found := obj.Maps[mapName]
	if !found {
		panic(fmt.Sprintf("Failed to find ring buffer map named '%s'", mapName))
	}

	rbReader, err := ringbuf.NewReader(ringBuf)
	if err != nil {
		panic(fmt.Sprintf("Failed to create ring buffer reader: %v", err))
	}
	defer rbReader.Close()


	signalChan := make(chan os.Signal, 1)
	signal.Notify(signalChan, os.Interrupt, syscall.SIGTERM)
	go func() {
		<-signalChan
		fmt.Println("\nExiting...")
		rbReader.Close()
		lnk.Close()
		os.Exit(0)
	}()


	for {
		record, err := rbReader.Read()
		if err != nil {
			if err == ringbuf.ErrClosed {
				break
			}
			//log.Printf("Error reading ring buffer: %v", err)
			continue
		}

		var evt PacketEvt
		reader := bytes.NewReader(record.RawSample)
		err = binary.Read(reader, binary.LittleEndian, &evt)
    	if err != nil {
        	log.Printf("Failed to decode event: %v", err)
        	continue
    	}
		//events := make(map[string]string)
		//events[formatIpPort(evt.SrcIP, evt.SrcPort)] = formatIpPort(evt.DestIP, evt.DestPort)
		//fmt.Printf("Received Event => (SRC) -> [%s] | (DEST) -> [%s]\n",
		//	formatIpPort(evt.SrcIP, evt.SrcPort), events[formatIpPort(evt.SrcIP, evt.SrcPort)])

		var events []Event
		var counter int
		counter = 0
		currentTime := time.Now()
		formattedTime := currentTime.Format("2006-01-02 15:04:05")
		events = append(events, Event{Source: formatIpPort(evt.SrcIP, evt.SrcPort), Destination: formatIpPort(evt.DestIP, evt.DestPort), Timestamp: formattedTime});
		fmt.Println(events[counter]);
		counter++;
	}
}
